### PR TITLE
[move core modules] add ApprovedPayment module

### PR DESCRIPTION
**Pull request by sblackshear**
New core module that allows a payee to approve payments with a cryptographic signature. The basic flow is:
(1) Payer sends payment metadata to the payee
(2) Payee signs metadata and sends 64 byte signature back to the payer
(3) Payer sends an approved payment to the payee via transaction invoking containing the payment metadata and a signature. The transaction will abort if the signature check fails.

Note: approved payment's are an accounting convenience/a courtesy mechanism for the payee, *not* a hurdle that must be cleared for all payments to the payee. In addition, approved payments do not have replay protection (that is, the approval signature can be replayed).

In addition to the core module, added scripts for registering an `ApprovedPayment` resource + wrappers for encoding in Rust.

## Motivation

See above

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

New Move functional tests and script E2E tests.

### CODE REVIEW

File: language/stdlib/modules/approved_payment.move [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+module ApprovedPayment {
+    use 0x0::Libra;
+    use 0x0::LibraAccount;
+    use 0x0::Signature;
+    use 0x0::Transaction;
+    use 0x0::Vector;
+
+    // A resource to be published under the payee's account
+    resource struct T {
+        // 32 byte EdDSA public key whose counterpart must be used to sign the payment data
```
Comment (kchalkias) [inline language/stdlib/modules/approved_payment.move]: Let’s mention here “single” Ed25519 key. 
I guess we don’t want to give optionality as in auth_key, right?
Also, maybe mention “unlike auth_key which uses hashed public keys”

File: language/stdlib/modules/approved_payment.move [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// Module that allows a payee to approve payments with a cryptographic signature. The basic flow is:
+// (1) Payer sends `metadata` to the payee
+// (2) Payee signs `metadata` and sends 64 byte signature back to the payer
+// (3) Payer sends an approved payment to the payee by sending a transaction invoking `deposit`
+//     with payment data + signature. The transaction will abort if the signature check fails.
+// Note: approved payment's are an accounting convenience/a courtesy mechansim for the payee, *not*
```
Comment (kchalkias) [inline language/stdlib/modules/approved_payment.move]: *payments

File: language/stdlib/modules/approved_payment.move [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// a hurdle that must be cleared for all payments to the payee. In addition, approved payments do
+// not have replay protection.
```
Comment (kchalkias) [inline language/stdlib/modules/approved_payment.move]: I guess we decided to not include sender and/or amount and/or receiver, right?

File: language/stdlib/transaction_scripts/register_approved_payment.move [hunk 4]
```diff
@@ -0,0 +1,8 @@
+use 0x0::ApprovedPayment;
+
+// Publish a newly created `ApprovedPayment` resource under the sender's account with approval key
+// `public_key`.
+// Abprts if the sender already has a published `ApprovedPayment` resource.
```
Comment (kchalkias) [inline language/stdlib/transaction_scripts/register_approved_payment.move]: *Aborts

File: language/stdlib/modules/approved_payment.move [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            payee,
+            LibraAccount::withdraw_from_sender<Token>(amount),
+            metadata,
+            signature
+        )
+    }
+
+    // Rotate the key used to sign approved payments. This will invalidate any approved payments
+    // that are currently in flight
+    public fun rotate_key(approved_payment: &mut T, new_public_key: vector<u8>) {
```
Comment (kchalkias) [inline language/stdlib/modules/approved_payment.move]: Out of curiosity, what is happening if I don't have a registered key and try to rotate it?

File: language/stdlib/modules/approved_payment.move (hunk 3 above)
Comment (sblackshear) [reply language/stdlib/modules/approved_payment.move]: Correct, although we haven't ruled out signing `amount` in the future. Couple of thoughts on why not sender/receiver:

- Signing the sender address limits the flexibility of the payer to initiate payment from any of multiple addresses they may control
- Even though the receiver address is not signed, this approach ensures that an approved payment to address *A* must use the `ApprovedPayment::T` resource published under *A*'s address. That is: you can't do an approved payment to *A1* using an approval signature from *A2* (assuming they use different approval keys).

File: language/stdlib/modules/approved_payment.move (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/modules/approved_payment.move]: Good question on optionality. It seems simpler to stick with just Ed25519, but would definitely be easy to add flexibility if there's a need for multiple schemes.

File: language/stdlib/modules/approved_payment.move (hunk 5 above)
Comment (sblackshear) [reply language/stdlib/modules/approved_payment.move]: `borrow_global<T>(a)` and `move_from<T>(a)` abort if there is no resource of type `T` published under address `a`. Similarly, `move_to_sender<T>` aborts if there is no resource of type `T` published under the sender's address. Basically, there's an implicit `exists` check in each case that will trigger an abort. 

We could choose to do an explicit `exists` check here if we wanted to abort with a more constructive error code. The cases above will abort with something fairly generic like `MISSING_DATA`. The circumstances under which you want to add such error codes is a topic of active debate in our team--it's a gas/code size vs error code clarity tradeoff.

File: language/move-vm/types/src/native_functions/signature.rs [hunk 6]
```diff
@@ -62,7 +62,8 @@ pub fn native_ed25519_signature_verification(
 
     let sig = match ed25519::Ed25519Signature::try_from(signature.as_slice()) {
         Ok(sig) => sig,
-        Err(_) => {
+        Err(e) => {
+            println!("failing with {:?}", e);
```
Comment (tzakian) [inline language/move-vm/types/src/native_functions/signature.rs]: Forgot to remove these `println`s?

File: language/stdlib/modules/approved_payment.move [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        // LibraAccount::T, which is a hash of a public key + signature scheme identifier.
+        public_key: vector<u8>,
+        // TODO: events?
+    }
+
+    // Deposit `coin` in `payee`'s account if the `signature` on the payment metadata matches the
+    // public key stored in the `approved_payment` resource
+    public fun deposit<Token>(
+        approved_payment: &T,
+        payee: address,
```
Comment (tzakian) [inline language/stdlib/modules/approved_payment.move]: Super nit (feel free to ignore): Other deposit functions in `LibraAccount` usually take the payee address as the first argument and the coin as second, followed by everything else. Was wondering if we wanted to also keep that convention here?

File: language/stdlib/modules/approved_payment.move [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+                *&approved_payment.public_key,
+                copy metadata
+            ),
+            9002, // TODO: proper error code
+        );
+        LibraAccount::deposit_with_metadata<Token>(payee, coin, metadata)
+    }
+
+    // Wrapper of `deposit` that withdraw's from the sender's balance and uses the top-level
+    // `ApprovedPayment.T` resource under the payee account.
```
Comment (tzakian) [inline language/stdlib/modules/approved_payment.move]: Nit: move syntax for the type.

File: language/move-vm/types/src/native_functions/signature.rs (hunk 6 above)
Comment (sblackshear) [reply language/move-vm/types/src/native_functions/signature.rs]: Derp, thanks!

File: language/stdlib/modules/approved_payment.move (hunk 7 above)
Comment (sblackshear) [reply language/stdlib/modules/approved_payment.move]: Good question--I was actually basing this on some code in `Libra` that implements the "support doing X on both a top-level resource and a resource stored in a field" pattern. That is:

```
do_X(t: &mut T, args) { // actually do X }
do_X_from_sender(args) { do_X(borrow_global_mut<T>(Transaction::sender(), args) }
```

I think for this pattern, it makes sense to put `&T` first and the args second to make it clear how the signatures are different.

File: language/stdlib/modules/approved_payment.move (hunk 8 above)
Comment (sblackshear) [reply language/stdlib/modules/approved_payment.move]: I keep doing this. Eventually I'll forget the IR...

File: language/stdlib/modules/approved_payment.move (hunk 1 above)
Comment (kchalkias) [reply language/stdlib/modules/approved_payment.move]: Optionality is not required for now. I guess the main question will always be "why raw PK here Vs hashing at the auth_key layer".

**Pull request closed without merging**